hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/upgrade/OMPrepareRequest.java

```diff
@@ -0,0 +1,196 @@
+      // so they will not be purged from the log before being persisted to
+      // the DB.
+      // Since the response for this request was added to the double buffer
+      // already, once this index reaches the state machine, we know all
+      // transactions have been flushed.
+      waitForDoubleBufferFlush(ozoneManager, transactionLogIndex);
+
+      OzoneManagerRatisServer omRatisServer = ozoneManager.getOmRatisServer();
+      RaftServerProxy server = (RaftServerProxy) omRatisServer.getServer();
+      RaftServerImpl serverImpl =
+          server.getImpl(omRatisServer.getRaftGroup().getGroupId());
+
+      takeSnapshotAndPurgeLogs(serverImpl);
+
+      // TODO: Create marker file with txn index.
+
+      LOG.info("OM prepared at log index {}. Returning response {}",
+          ozoneManager.getRatisSnapshotIndex(), omResponse);
+    } catch (IOException e) {
+      response = new OMPrepareResponse(
+          createErrorOMResponse(responseBuilder, e));
+    } catch (InterruptedException e) {
+      response = new OMPrepareResponse(
+          createErrorOMResponse(responseBuilder, new OMException(e,
+              OMException.ResultCodes.INTERNAL_ERROR)));
+    }
+
+    return response;
+  }
+
+  private static void waitForDoubleBufferFlush(
+      OzoneManager ozoneManager, long txnLogIndex)
+      throws InterruptedException, IOException {
+
+    long endTime = System.currentTimeMillis() +
+        DOUBLE_BUFFER_FLUSH_TIMEOUT.toMillis();
+    boolean success = false;
+
+    while (!success && System.currentTimeMillis() < endTime) {
+      // If no transactions have been persisted to the DB, transaction info
+      // will be null, not zero, causing a null pointer exception within
+      // ozoneManager#getRatisSnaphotIndex.
+      // Get the transaction directly instead.
+      OMTransactionInfo txnInfo = ozoneManager.getMetadataManager()
+          .getTransactionInfoTable().get(TRANSACTION_INFO_KEY);
+      if (txnInfo == null) {
+        success = (txnLogIndex == 0);
+      } else {
+        success = (txnInfo.getTransactionIndex() == txnLogIndex);
+      }
+
+      Thread.sleep(DOUBLE_BUFFER_FLUSH_CHECK_INTERVAL.toMillis());
+    }
+
+    if (!success) {
+      throw new IOException(String.format("After waiting for %d seconds, " +
+              "State Machine has not applied  all the transactions.",
+          DOUBLE_BUFFER_FLUSH_TIMEOUT.toMillis() * 1000));
+    }
+  }
+
+  /**
+   * Take a snapshot of the state machine at the last index, and purge ALL logs.
+   * @param impl RaftServerImpl instance
+   * @throws IOException on Error.
+   */
+  public static long takeSnapshotAndPurgeLogs(RaftServerImpl impl)
+      throws IOException {
+
+    StateMachine stateMachine = impl.getStateMachine();
+    long snapshotIndex = stateMachine.takeSnapshot();
+    if (snapshotIndex != stateMachine.getLastAppliedTermIndex().getIndex()) {
+      throw new IOException("Index from Snapshot does not match last applied " +
```

Comment:
Can we add a comment here: What is the recovery step from this? If the snapshot index does not match will subsequent prepare ever succeed? cc: @avijayanhwx

Reply:
If this fails, prepare will fail and an error response will be sent back to the client. I will add a comment to address this.